src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormatter.cs

```diff
@@ -88,5 +92,46 @@ private TextEdit[] MapEditsToHostDocument(RazorCodeDocument codeDocument, TextEd
 
             return actualEdits.ToArray();
         }
+
+        private async Task<TextEdit[]> FormatOnClient(RazorCodeDocument codeDocument,
+            Range projectedRange,
+            Uri uri,
+            FormattingOptions options)
+        {
+            var @params = new RazorDocumentRangeFormattingParams()
+            {
+                Kind = RazorLanguageKind.CSharp,
+                ProjectedRange = projectedRange,
+                HostDocumentFilePath = _filePathNormalizer.Normalize(uri.GetAbsoluteOrUNCPath()),
+                Options = options
+            };
+
+            var result = await _server.Client.SendRequest<RazorDocumentRangeFormattingParams, RazorDocumentRangeFormattingResponse>(
+                LanguageServerConstants.RazorRangeFormattingEndpoint, @params);
+
+            return result.Edits;
+        }
+
+        private async Task<TextEdit[]> FormatOnServer(RazorCodeDocument codeDocument,
+            Range projectedRange,
+            Uri uri,
+            FormattingOptions options)
+        {
+            var workspace = new AdhocWorkspace();
```

Comment:
We should have a singular workspace so we don't re-instantiate this each time. I'm pretty sure it's not cheap.